### DISCUSSION TITLE
Assign arbitrary colours to raw channels

### DISCUSSION THREAD

**Discussion by LAK132** [Beep6581/RawTherapee#6914] [Ideas]
Perhaps vaguely related to #5849, a full spectrum converted camera with a 550nm long pass filter is arguably IR+R/IR+G/IR+G/IR rather than RGGB, and I also have a camera that is IRRGB, so it would be nice to be able to correct for this manually before any of the other tools can misinterpret the data (such as chromatic aberration correction and white balance). I've tried using the channel mixer with a 550nm long pass camera, but because it happens so late in the processing chain it makes it impossible to use most other tools with it. 
I'd be happy to work on this, and I've tried looking at the code, but I'm not entirely sure where raw sensor data processing code goes.

**Comment by Lawrence37** [comment]
4 color raw images may require special handling beyond just manipulating the colors (demosaicing algorithm for example). If you're talking about 3 color images, maybe you can create a custom color profile. If you want it to be easily adjustable, then maybe what you want is the ability to manually enter the color matrix. For that, you can refer to `rtgui/icmpanel.cc` for the color management GUI and the `rgbProc` function in `rtengine/improcfun.cc` for where the color profile (DCP) is applied or search for `rgb_cam` for where the camera color matrix is applied.

**Comment by LAK132** [comment]
I ended up trying to make a custom profile for the camera (input profile) with the 3 primaries rotated around, however RT doesn't seem to be respecting this and the image just gets darker. Not sure if I'm doing something wrong here

**Comment by LAK132** [comment]
... looks like the issue I was having is that RT cached a bad version of the profile and attempting to load the profile again didn't invalidate the cache

**Comment by Lawrence37** [comment]
I wasn't aware of any profile caching done by RawTherapee. Were you able to load the correct profile, and does it accomplish what you initially wanted?

**Comment by LAK132** [comment]
a quick glance at the code made it look like it was caching based on file name. and yeah, if I reload RT it was able to load a working version of the profile which switched all of the primaries around. it doesn't compensate for IR ending up on all 3 channels but it does give me a much better starting point. I'll investigate profiles more and see if they're enough to compensate or if RT will still need explicit support for channel mixing prior to white balance and such

**Comment by LAK132** [comment]
I've taken another crack at this by attempting to do the IR channel subtraction immediately after demosaicing is done, however I am running into the issue of a bunch of preprocessing that happens before demosaicing seemingly messing with values. Is there some sane way to do consistent inter-channel manipulations prior to demosacing? Preferably without having to specialise on all sensor layouts?
